### DISCUSSION TITLE
iPXE No Network Device Found

### DISCUSSION THREAD

**Discussion by 266-750Balloons** [ltsp/ltsp#738] [Q&A]
I have a single AMD64 Debian Bookworm Server and a Pentium 4 laptop with 2 GB RAM as a client (mainly because I need to backup an old IDE drive, and this is the hardware I have on hand). I went through the tutorial and after opening some firewall ports, iPXE successfully starts up and seemingly gets an address assigned by the router, only to output the following errors:
![image](https://user-images.githubusercontent.com/79549730/184519812-c6490edf-5fff-4a02-975b-9424faaac84a.png)

When running `` dhcp `` in the iPXE console, it works. However, `` echo ${filename} `` and `` echo ${root-path} `` output nothing, suggesting that my server does not send those.

When running `` journalctl -fu dnsmasq ``, nothing shows when the device is booting. I've tried every seemingly related port I can find. The IP for my server is 192.168.0.152. I have no idea where the 192.168.0.148 is coming from unless it is the address assigned to the laptop.

DIstro: Debian Bookworm
LTSP version: 22.01-2
iPXE version: 1.0.0+git-20198125.36a4c85-5.1, 1.21.1 (I upgraded it to see if it was an issue with the Debian package's mega-outdated version. It did load, but with the same error._
Client Network Card: Intel PRO/100 VE

Thank you for your time.

**Comment by alkisg** [comment]
You have a typo, `sudo journalctl -fu dnsmasq` with q instead of w. Run it properly and upload the output.

It sounds like you didn't run the `sudo ltsp ipxe` command.

**Comment by 266-750Balloons** [comment]
Whoops. I made a typo in the description. I did type dnsmasq originally in the terminal. My mistake was, in fact, that I don't run the command as the superuser. Thanks for that correction. Here's the output.
```
Aug 13 19:40:47 drewbiac2600 dnsmasq-dhcp[9675]: PXE(enp6s0) 00:08:0d:db:47:fc proxy
Aug 13 19:40:49 drewbiac2600 dnsmasq-dhcp[9675]: PXE(enp6s0) 192.168.0.148 00:08:0d:db:47:fc ltsp/undionly.kpxe
Aug 13 19:40:50 drewbiac2600 dnsmasq-tftp[9675]: error 0 TFTP Aborted received from 192.168.0.148
Aug 13 19:40:50 drewbiac2600 dnsmasq-tftp[9675]: sent /srv/tftp/ltsp/undionly.kpxe to 192.168.0.148
Aug 13 19:40:50 drewbiac2600 dnsmasq-tftp[9675]: sent /srv/tftp/ltsp/undionly.kpxe to 192.168.0.148
Aug 13 23:47:16 drewbiac2600 dnsmasq-dhcp[9675]: PXE(enp6s0) 00:08:0d:db:47:fc proxy
Aug 13 23:47:18 drewbiac2600 dnsmasq-dhcp[9675]: PXE(enp6s0) 192.168.0.168 00:08:0d:db:47:fc ltsp/undionly.kpxe
Aug 13 23:47:19 drewbiac2600 dnsmasq-tftp[9675]: error 0 TFTP Aborted received from 192.168.0.168
Aug 13 23:47:19 drewbiac2600 dnsmasq-tftp[9675]: sent /srv/tftp/ltsp/undionly.kpxe to 192.168.0.168
Aug 13 23:47:19 drewbiac2600 dnsmasq-tftp[9675]: sent /srv/tftp/ltsp/undionly.kpxe to 192.168.0.168
```
Thank you for responding so quickly.

**Comment by alkisg** [comment]
Try to follow this page: https://ltsp.org/guides/pxelinux/
...but ONLY up to this line:

> Thus, the rest of this page refers to the following scenario:

Do NOT follow the rest of the page. Does it work after running the sudo wget command?

**Comment by 266-750Balloons** [comment]
It seems to have failed despite the step.
![374043E3-C935-46C7-B586-1BCF6BE0B553](https://user-images.githubusercontent.com/79549730/184527813-ca48b279-6e7d-4bd6-b833-710086cd23b6.jpeg)
Disregard the net0 one. I think that's the wifi card, and I'm using Ethernet. net1 is probably the one I am using (the Ethernet) since its IP has been the one showing in the logs (earlier, it was the 148 address).

When I get I chance, I'll probably try the PXELinux measures.

**Comment by alkisg** [comment]
I'm not sure the problem is in ipxe, it might be in dhcp, as you're not seeing a line like this in your logs:

```syslog
Jan 13 22:30:40 alkis dnsmasq-tftp[1269]: sent /srv/tftp/ltsp/ltsp.ipxe to 192.168.67.101
```

Try to press Ctrl+B right before the "No more network devices" step, and then type `config` to enter the menus, and see what filename the proxydhcp/ "directory" has.
If possible, also try with a different client (e.g. a VM) just to make sure that your server is set up correctly.

**Comment by 266-750Balloons** [comment]
I'm not even seeing `` proxdhcp/ `` within the menu. When I have time, I'll try to test with a VM when I have time, but I think in order to properly test, I need to set up a tap network for qemu.
In case it's important, I just have dhcp on defaults; e.g IPs are being assigned by the router.

**Comment by 266-750Balloons** [comment]
I tested a newer laptop (Vista-era machine) to receive the exact same error. I've tried messing round with ltsp-dnsmasq.conf to no avail. It is probably a DHCP issue, but I am no expert on the configs. After messing around, here is what I have in the way of configs:
* /etc/dnsmasq.d/ltsp-dnsmasq.conf
```
# This file is part of LTSP, https://ltsp.org
# Copyright 2019-2021 the LTSP team, see AUTHORS
# SPDX-License-Identifier: GPL-3.0-or-later

# Configure dnsmasq for LTSP
# Documentation=man:ltsp-dnsmasq(8)

# For additional local dnsmasq configuration like DNS blacklisting, it's
# recommended to use separate /etc/dnsmasq.d/local.conf files,
# so that they're not lost if you ever (re)run `ltsp dnsmasq`.

# port=0 disables the DNS service of dnsmasq
port=0

# enable-tftp enables the TFTP service of dnsmasq
enable-tftp

# FHS 2.3+ recommends /srv for tftp (debian #477109, LP #84615)
tftp-root=/srv/tftp

# Log lots of extra information about DHCP transactions
#log-dhcp

# IP ranges to hand out, usually on the internal LTSP subnet of 2-NIC setups
dhcp-range=192.168.67.20,192.168.67.250,12h

# If another DHCP server is present on the network, a proxy range may be used
# instead. This makes dnsmasq provide boot information but not IP leases.
dhcp-range=set:proxy,192.168.0.152,proxy,255.255.255.0

# Specify the DNS server. 0.0.0.0 means the machine running dnsmasq.
# DNS_SERVER in ltsp.conf is preferred as it reaches proxy DHCP clients.
dhcp-option=option:dns-server,192.168.0.1,192.168.0.1

# Set some tags to be able to separate client settings later on.
dhcp-match=set:X86PC,option:client-arch,0
dhcp-match=set:X86-64_EFI,option:client-arch,7
# Due to rfc4578 errata, sometimes BC_EFI=9 is misused instead of X86-64_EFI=7:
dhcp-match=set:X86-64_EFI,option:client-arch,9
dhcp-mac=set:rpi,b8:27:eb:*:*:*
dhcp-mac=set:rpi,dc:a6:32:*:*:*
dhcp-mac=set:rpi,e4:5f:01:*:*:*

# Check if iPXE has the features we need: http://ipxe.org/howto/dhcpd
# http && menu && ((pxe && bzimage) || efi)
dhcp-match=set:ipxe-http,175,19
dhcp-match=set:ipxe-menu,175,39
dhcp-match=set:ipxe-pxe,175,33
dhcp-match=set:ipxe-bzimage,175,24
dhcp-match=set:ipxe-efi,175,36
tag-if=set:ipxe,tag:ipxe-http,tag:ipxe-menu,tag:ipxe-pxe,tag:ipxe-bzimage
tag-if=set:ipxe,tag:ipxe-http,tag:ipxe-menu,tag:ipxe-efi

# In proxy DHCP mode, the server ONLY sends its IP and the following filename.
# Service types: man dnsmasq or https://tools.ietf.org/html/rfc4578#section-2.1
# PXE services in non proxy subnets sometimes break UEFI netboot, so tag:proxy.
pxe-service=tag:proxy,tag:!ipxe,X86PC,"undionly.kpxe",ltsp/undionly.kpxe
pxe-service=tag:proxy,tag:!ipxe,X86-64_EFI,"snponly.efi",ltsp/snponly.efi
pxe-service=tag:proxy,tag:ipxe,X86PC,"ltsp.ipxe",ltsp/ltsp.ipxe
pxe-service=tag:proxy,tag:ipxe,X86-64_EFI,"ltsp.ipxe",ltsp/ltsp.ipxe
pxe-service=tag:rpi,X86PC,"Raspberry Pi Boot   ",unused

# Specify the boot filename for each tag, relative to tftp-root.
# If multiple lines with tags match, the last one is used.
# See: https://www.syslinux.org/wiki/index.php?title=PXELINUX#UEFI
dhcp-boot=tag:!ipxe,tag:X86PC,ltsp/undionly.kpxe
dhcp-boot=tag:!ipxe,tag:X86-64_EFI,ltsp/snponly.efi
dhcp-boot=tag:ipxe,ltsp/ltsp.ipxe

dhcp-no-override

# Proxy DHCP clients don't receive any DHCP options like root-path.
# So we set root-path in the kernel cmdline from ltsp.ipxe.
# dhcp-option=option:root-path,ipxe-menu-item

```
* /etc/dnsmasq.conf
```
dhcp-no-override
tftp-no-blocksize
```
Do you see anything weird?

**Comment by 266-750Balloons** [comment]
I am totally stumped at this point.

**Comment by alkisg** [comment]
If you want, you may come from IRC/Matrix for live support.
Do have VirtualBox installed so that we can test with a VM client.

**Comment by 266-750Balloons** [comment]
Rather unfortunately, I can't due to major time zone differences and schedule constraints (I'm UTC-7).
I can install VirtualBox in the next few days. (I just haven't installed it since I switched off Windows because Virt-Manager and QEMU-KVM have been enough so far. However, I do know of some of VirtualBox's networking convenience.)
If it's too difficult to do any other diagnostics without live support, then it's fine if this question is never answered. It's nothing pressing, and I'm sure you have more serious matters to attend to. Thank you for your help and support so far.

**Comment by alkisg** [comment]
Virt-Manager should also be fine for testing the client side, but unfortunately I have no experience with it.
UTC-7 mornings match UTC+3 afternoons, that's not a big issue.
You may also seek help from the iPXE community, since this is more related to their code base than the LTSP one.
